packages/state-transition/src/cache/epochCache.ts

```diff
@@ -748,9 +771,74 @@ export class EpochCache {
     return isAggregatorFromCommitteeLength(committee.length, slotSignature);
   }
 
+  /**
+   * Return finalized pubkey given the validator index.
+   * Only finalized pubkey as we do not store unfinalized pubkey because no where in the spec has a
+   * need to make such enquiry
+   *
+   */
+  getPubkey(index: ValidatorIndex): PublicKey {
+    return this.index2pubkey[index];
+  }
+
+  getValidatorIndex(pubkey: Uint8Array | PubkeyHex): ValidatorIndex | undefined {
+    if (this.isAfterEIP6110()) {
+      return this.pubkey2index.get(pubkey) ?? this.unfinalizedPubkey2index.get(toMemoryEfficientHexStr(pubkey));
+    } else {
+      return this.pubkey2index.get(pubkey);
+    }
+  }
+
+  /**
+   *
+   * Add unfinalized pubkeys
+   *
+   */
   addPubkey(index: ValidatorIndex, pubkey: Uint8Array): void {
+    if (this.isAfterEIP6110()) {
+      this.unfinalizedPubkey2index = this.unfinalizedPubkey2index.set(toMemoryEfficientHexStr(pubkey), index);
+    } else {
+      this.pubkey2index.set(pubkey, index);
+      this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian);
+    }
+  }
+
+  addFinalizedPubkeys(pubkeyMap: UnfinalizedPubkeyIndexMap): void {
+    Array.from(pubkeyMap.entries()).forEach(([pubkey, index]) => this.addFinalizedPubkey(index, pubkey));
```

Comment:
For iterate and for loop, no need to convert to array first and loop latter with foreach